Services/Products/ProductsService.php

```diff
@@ -0,0 +1,287 @@
+				'unitary_value' => $product->unitary_value,
+				'weight'        => $product->weight,
+				'width'         => $product->width,
+				'height'        => $product->height,
+				'length'        => $product->length,
+			);
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Function to filter products to api Melhor Envio.
+	 *
+	 * @param array $products
+	 * @return array
+	 */
+	public function filter( $data ) {
+		$products = array();
+		foreach ( $data as $key => $item ) {
+			if ( ! empty( $item->shipping_fee ) && $item->shipping_fee == 'whole' ) {
+				$item->components = [];
+			}
+
+			if ( ! empty( $item->shipping_fee ) && $item->shipping_fee == 'each' && ! empty( $item->components ) ) {
+				foreach ($item->components as $component) {
+					$products[] = $component;
+				}
+				continue;
+			}
+
+			if ( $this->isObjectProduct( $item ) ) {
+				$data       = $item->get_data();
+				$product    = $item;
+				$products[$key] = $this->normalize(
+					$product,
+					$product->get_price(),
+					$item['quantity']
+				);
+				continue;
+			}
+
+			if ( ! empty( $item->name ) && ! empty( $item->id ) ) {
+				$products[$key] = $item;
+				continue;
+			}
+
+			if ( ! empty( $item['name'] ) && ! empty( $item['id'] ) ) {
+				$products[$key] = (object) $item;
+				continue;
+			}
+
+			$product    = $item['data'];
+			$products[$key] = $this->normalize(
+				$product,
+				$product->get_price(),
+				$item['quantity']
+			);
+		}
+
+		return $products;
+	}
+
+	/**
+	 * @param object $product
+	 * @return bool
+	 */
+	private function isObjectProduct( $item ) {
+		return (
+			! is_array( $item ) &&
+			(
+				get_class( $item ) == WooCommerceBundleProductsService::OBJECT_PRODUCT_SIMPLE ||
+				get_class( $item ) == WooCommerceBundleProductsService::OBJECT_WOOCOMMERCE_BUNDLE
+			)
+		);
+	}
+
+	public function getDataByProductCart( $productCart , $items): Product
+	{
+		$data = self::normalize(
+			$productCart['data'],
+			$productCart['data']->get_price(),
+			$productCart['quantity']
+		);
+
+		if (isset($productCart['wooco_parent_id'])){
+			$data->parentId = $productCart['wooco_parent_id'];
+		}
+
+		if (isset($productCart['woosb_parent_id'])){
+			$data->parentId = $productCart['woosb_parent_id'];
+		}
+
+		return $data;
+	}
+
+	public function getDataByProductOrder( $productOrder, $items)
+	{
+		$data = self::normalize(
+			$productOrder->get_product(),
+			$productOrder->get_product()->get_price(),
+			$productOrder->get_quantity()
+		);
+
+		if (! empty( $productOrder->get_meta('wooco_parent_id', true) )){
+			$data->parentId = $productOrder->get_meta('wooco_parent_id', true);
+		}
+
+		if (! empty( $productOrder->get_meta('_woosb_parent_id', true) )){
+			$data->parentId = $productOrder->get_meta('_woosb_parent_id', true);
+		}
+
+		return $data;
+	}
+	/**
+	 * @param $product
+	 * @param $price
+	 * @param int $quantity
+	 * @return Product
+	 */
+	public function normalize($product, $price, $quantity = 1): Product
```

Comment:
Fiquei um pouco confuso com esse nome `$product`. No caso, isso seria um daqueles itens dentro do produto, o agregado de produtos, certo?

Um nome diferente de `product` não ficaria mais fácil de entender? Minha sugestão é `productAggregate` pra poder ficar diferente de `compositeProduct` ou `bundleProduct`, o que causaria confusão com os plugins.

```suggestion
	public function normalize($productAggregate, $price, $quantity = 1): Product
```

O que acha?

Reply:
Essa função `normalize()` é utilizada para todos os tipos de produtos, simples, composto, componente, qualquer tipo de produto pode ser passado ai, o objetivo dela é padronizar o objeto produto para um "formato" que julgamos mais fácil de lidar e com as informações que precisamos